docs/language.md

```diff
@@ -48,41 +52,106 @@ a `validate` expression.
 
 Built-in base types are also similar to JavaScript types:
 
-    String  - Stings
-    Number  - Integer or floating point
-    Boolean - Values true or false
-    Object  - A structured object containing named properties.
-    Any     - Every non-null value is of type Any.
-    Null    - Value null (same as absence of a value, or deleted)
+    String            - Stings
+    Number            - Integer or floating point
+    Boolean           - Values true or false
+    Object            - A structured object containing named properties.
+    Any               - Every non-null value is of type Any.
+    Null              - Value null (same as absence of a value, or deleted)
+    Map<Key, Value>   - A generic type - maps string valued keys to corresponding
+                        values (similar to an Object type).
+    Type[]            - An "array-like" type (actually same as Map<String, Type>
+                        where Type can be any other built-in or user-defined type.
 
-You can _extend_ any of the built-in scalar types by adding a validation expression, e.g.:
+Any of the built-in scalar types can be _extended_ by adding a validation expression, e.g.:
 
-    type ShortString extends String {
-      validate() = this.length < 32;
-    }
+```javascript
+type ShortString extends String {
+  validate() = this.length < 32;
+}
 
-    type Percentage extends Number {
-      validate() = this >=0 && this <= 100;
-    }
+type Percentage extends Number {
+  validate() = this >=0 && this <= 100;
+}
+```
+
+Notes:
+
+- Object types are required to have at least one property when present.
+- Map types can be empty collections (they need not contain any child keys).
 
 ## Type Expressions
 
-Any place a Type can be used, it can be replaced with a Type expression.
+Any place a Type can be used, it can be replaced with a Type expression.  We support
+three types of type expressions:
+
+### Union Types
 
     Type1 | Type2    - Value can be either of two types.
     Type | Null      - An optional `Type` value (value can be deleted or missing).
 
+
+### Map Types (Collections)
```

Comment:
Not really.  Collections are just Objects with unique keys for each member.

If you want to limit the max size for small collections you could do something like this (not that all keys are strings - not integers):
```
type SingleDigit extends String {
  validate() = this.length == 1 && this >= '0' && this <= '9';
}

type Group {
  members: Map<SingleDigit, User>
}
```

See this sample:

https://github.com/firebase/bolt/blob/koss-generics-2/samples/groups.bolt